### ISSUE TITLE
dry run option just to generate the report?

### ISSUE THREAD

**Issue by adrian-herscu** [TNG/JGiven#418]
Is there any option for running a dry run just to see how the test report would be generated?
Any suggestion about how to implement such thing?

**Comment by janschaefer** [comment]
The `@Pending` annotation could be used for single tests (https://static.javadoc.io/com.tngtech.jgiven/jgiven-core/0.18.0/com/tngtech/jgiven/annotation/Pending.html), but I am not sure if this is what you want.

**Comment by adrian-herscu** [comment]
> The `@Pending` annotation could be used for single tests (https://static.javadoc.io/com.tngtech.jgiven/jgiven-core/0.18.0/com/tngtech/jgiven/annotation/Pending.html), but I am not sure if this is what you want.

Yes, I am already using it for problematic tests.
I am looking for a central hooking point that I can use in order to make all stages temporarily appear as pending, just for one run. Thought about using AspectJ to add the pending annotation at runtime but still not found whether such thing as adding annotations at runtime is supported.

**Comment by janschaefer** [comment]
Ok, we could think about a global configuration setting for that. May I ask what your concrete use case is?

**Comment by adrian-herscu** [comment]
It's a bit complex... I am working for a medical device company. We want automated tests to run nightly for regression purpose. Jenkins+Maven+TestNG+JGiven are fine for that :) We publish the JGiven report and everybody is happy! On the other hand, we cannot prove that the automated tests are valid, i.e. that they do whatever they are supposed to do, and do that without substantial effort. Since these tests are custom made for our product, they change as new features are added/modified, so we would need to validate them for every release. In simple words, we need to show how we test our tests and also do those tests! This effort is definitely greater than just running these tests manually. Still, we need to issue a test document to be executed and signed by manual testers. We already implemented a reporter to do just that. It runs via Maven, gathers all JGiven's JSON files and spits out an HTML document that latter can be imported into Word for additional mingling. But in order to have those JSON files, we still need to actually run these tests. Besides consuming lots of time, it also fails sometimes, requiring a new run. All this just to issue a test document for manual execution!
If there would be a dry run switch, we could get all JSON report files completely and could run our report generator on them :) Hope it makes sense...

**Comment by janschaefer** [comment]
Ok, I think I understand. I also think that it is not difficult to implement, given that this mainly means to execute all scenarios as if they are pending.

**Comment by janschaefer** [comment]
From the implementation point of view, I would say to introduce a new global configuration option that can be set via a system property. This has to be evaluated by the test runner and passed to the scenario executor.

**Comment by adrian-herscu** [comment]
> From the implementation point of view, I would say to introduce a new global configuration option that can be set via a system property. This has to be evaluated by the test runner and passed to the scenario executor.

That would suffice; meanwhile, can you please suggest a way of implementing that behavior from outside?

**Comment by adrian-herscu** [comment]
Here is an AspectJ class that implements dry-run mode:
```
import static org.apache.commons.lang3.StringUtils.*;

import org.aspectj.lang.*;
import org.aspectj.lang.annotation.*;

import com.tngtech.jgiven.annotation.*;

import lombok.extern.slf4j.*;

/**
 * Monitors execution of JGiven step methods. If {@code dryrun} system property
 * is not blank then it shortcuts the actual method execution.
 */
@Aspect
@Slf4j
public class DryRunAspect {
    public static final boolean dryRun =
        isNotBlank(System.getProperty("dryrun"));

    /**
     * Prevents execution if in {@code dryrun} mode.
     * 
     * @param thisJoinPoint
     *            advised join point
     * @return forwarded from advised method
     * @throws Throwable
     *             re-thrown as received advised method
     * @see #stageMethod()
     * @see #stepMethod()
     */
    @Around("stageMethod() && stepMethod()")
    @SuppressWarnings("static-method")
    public Object aroundStepMethod(
        final ProceedingJoinPoint thisJoinPoint)
        throws Throwable {
        if (dryRun) {
            log.trace("dryrun -- skipping {}:{}",
                thisJoinPoint.getTarget(),
                thisJoinPoint.getSignature().getName());
            return thisJoinPoint.getTarget();
        }

        return thisJoinPoint.proceed(thisJoinPoint.getArgs());
    }

    /**
     * Prevents execution if in {@code dryrun} mode.
     *
     * @param thisJoinPoint
     *            advised join point
     * @return forwarded from advised method
     * @throws Throwable
     *             re-thrown as received advised method
     * @see #stageMethod()
     * @see #stepMethod()
     */
    @Around("notUnitilsScenarioTestMethod() && beforeOrAfterMethod()")
    @SuppressWarnings("static-method")
    public Object aroundBeforeOrAfterMethod(
        final ProceedingJoinPoint thisJoinPoint)
        throws Throwable {
        // FIXME duplicated from above
        if (dryRun) {
            log.trace("dryrun -- skipping {}:{}",
                thisJoinPoint.getTarget(),
                thisJoinPoint.getSignature().getName());
            return thisJoinPoint.getTarget();
        }

        return thisJoinPoint.proceed(thisJoinPoint.getArgs());
    }

    /**
     * Matches the execution of any method within a sub-type of
     * {@link com.tngtech.jgiven.base.StageBase}.
     */
    @Pointcut("within(com.tngtech.jgiven.base.StageBase+)")
    public void stageMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of any method returning SELF (a subclass of
     * {@link com.tngtech.jgiven.base.StageBase}.
     */
    @Pointcut("execution(com.tngtech.jgiven.base.StageBase+ *(..))")
    public void stepMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of any method not within
     * {@link com.ge.acutecare.qa.jgiven.commons.utils.UnitilsScenarioTest}.
     */
    @Pointcut("within(!com.ge.acutecare.qa.jgiven.commons.utils.UnitilsScenarioTest)")
    public void notUnitilsScenarioTestMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with
     * {@link org.testng.annotations.BeforeMethod}.
     */
    @Pointcut("execution("
        + "@org.testng.annotations.BeforeMethod "
        + "* *(..))")
    public void beforeMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with
     * {@link org.testng.annotations.AfterMethod}.
     */
    @Pointcut("execution("
        + "@org.testng.annotations.AfterMethod "
        + "* *(..))")
    public void afterMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with
     * {@link org.testng.annotations.BeforeClass}.
     */
    @Pointcut("execution("
        + "@org.testng.annotations.BeforeClass "
        + "* *(..))")
    public void beforeClass() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with
     * {@link org.testng.annotations.AfterClass}.
     */
    @Pointcut("execution("
        + "@org.testng.annotations.AfterClass "
        + "* *(..))")
    public void afterClass() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with {@link BeforeStage}.
     */
    @Pointcut("execution("
        + "@com.tngtech.jgiven.annotation.BeforeStage "
        + "* *(..))")
    public void beforeStage() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with {@link AfterStage}.
     */
    @Pointcut("execution("
        + "@com.tngtech.jgiven.annotation.AfterStage "
        + "* *(..))")
    public void afterStage() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with {@link BeforeScenario}.
     */
    @Pointcut("execution("
        + "@com.tngtech.jgiven.annotation.BeforeScenario "
        + "* *(..))")
    public void beforeScenario() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of methods annotated with {@link AfterScenario}.
     */
    @Pointcut("execution("
        + "@com.tngtech.jgiven.annotation.AfterScenario "
        + "* *(..))")
    public void afterScenario() {
        // nothing to do here -- just defines a pointcut matcher
    }

    /**
     * Matches the execution of before or after methods.
     */
    @Pointcut("beforeMethod() || afterMethod()"
        + " || beforeClass() || afterClass()"
        + " || beforeStage() || afterStage()"
        + " || beforeScenario() || afterScenario()")
    public void beforeOrAfterMethod() {
        // nothing to do here -- just defines a pointcut matcher
    }
}

```

**Comment by janschaefer** [comment]
Fixed by #435
